### DISCUSSION TITLE
ArcGIS Enterprise Portal

### DISCUSSION THREAD

**Discussion by roimartins** [R-ArcGIS/arcgislayers#154] [Q&A]
Hi, thanks for the package creation . this will provide capabilities to promote ArcGIS Enterprise as a spatial data server for any data science solution and not only linked to users with an ArcGIS Pro installation ( as happens with arcgisbinding) . 

At the moment I managed to obtain n httr2 token using auth_user() fuinction with my built-in Portal user with user name and password. However this token is not working when used with with arc_open() function: 

```r
> arcgislayers::arc_open( "https://name_of_the_server/server/rest/services/Hosted/name_of_the_feature/FeatureServer/0", token = user_token ) 
Error in `arcgislayers::arc_open()`:
! Status code: 498
Error : Invalid Token
```

Then I tried to use the other authentication methods with my user with OAuth approach but I didnt find the instructions to follow this approach up . Unfortunaley in the README file this link looks it is broken :  see the `{arcgislayers}` [authorization article](articles/Authorization.html).

I have tried with auth_client() function but  this requires a ArcGIS Developer accoutn linked to the AGOL. However, it is not possible to link ArcGIS Developer account to an ArcGIS Enterprise Portal , only gives the option to use a AGOL URL . 

 
Unfortunately , so far the only method to  connect to not public layers in AGOL / AGE  is arcgisbinding , with the limitation above mentioned.

**Comment by JosiahParry** [comment]
Thanks, @roimartins. I will work on removing that link from the README. Please see this article: https://r.esri.com/r-bridge-site/location-services/connecting-to-a-portal.html

Per the doc, the recommendation is to use `auth_code()`. You should avoid using `auth_user()` as that requires a username and password which is a security risk.

**Reply by roimartins** [reply]
Fantastic  @JosiahParry , it works perfectly fine following the instructions with ArcGIS Enterprise ( even with AGE version 10.9.1 but you have to go 'settings -> Application -> application Registration ' and it is 'App Id' and 'App Secret') . 

So , other users registered in AGE Portal can use the same App Id and Secret for connect , but I guess they have to use their username instead.  Do I have to share the 'App' item with the 'organisation' for any other user in  organisation have access to this item and therefore use for R connection? 


In regard to 'auth_user()'  I didnt manage to obtain a valid token until I get your httr2:Oauth method into a R function I created on my own but using httr. I think the key is the 'client' and 'refere' parameter. In your function is set with 'requestip' and not sure if due to proxy , etc. the toke result of 'auth_user()' gave an error   until i use the function below.  Then I was able to use the result token in arc_open() function.  

```r
getAGEToken = function ( ) {

  portalUrl = "your_portal_url"

  request_body = list(  username =  "built_in_portal_user_name",                                
                        password = "user_password",                              
                        client=  'referer',    #'requestip',#'referer' ,
                        referer = portalUrl,
                        expiration=60 ,
                        f = 'pjson'
  )

  result_post = httr::POST(url  = "your_portal_url/sharing/rest/generateToken/",
                           body = request_body,
                           headers = c('content-type' = 'application/x-www-form-urlencoded')
                           )

 
  result_text = httr::content(result_post,  as="text")
  result =  jsonlite::fromJSON (result_text)

return  (
  httr2::oauth_token(result[["token"]], expires_in = 60,
                     username = "your_user_name", arcgis_host = portalUrl  )
)


  #return ( list (result_post,  result_text, result$token ) )

}
```


Anyhow better use OAuth instead, amazing job, thanks again, this approach will save a lots of future headaches and wider usability of AGE data services.

**Reply by roimartins** [reply]
hi , 'arc_open'  get access to the feature correctly . It is only when the result is used in arc_select that rise this error: 

```r

> arc_select(feature1 )
Error: 'req_perform_parallel' is not an exported object from 'namespace:httr2'

```

**Reply by JosiahParry** [reply]
Thanks for sharing your version of the auth token function. I'll see if swapping to `referrer` will fix it. 

When you installed `arcgisutils` and `arcgislayers` were you not prompted to install `httr2` version 1.0.0 or greater? They are hard requirements.

**Reply by roimartins** [reply]
I didnt have installed laterst httr2 version to include 'req_perform_parallel' function . Now works fine ! :)

**Reply by roimartins** [reply]
I think I wasnt prompted. But I have been doing many test last hours , so may it was and I ignored.

**Reply by JosiahParry** [reply]
Thanks, @roimartins! But when you installed the package, did it ask you to install it? It's listed as a hard dependency here https://github.com/R-ArcGIS/arcgislayers/blob/a823785c153487da707c3227f1c7ad8c12353b93/DESCRIPTION#L22

**Reply by JosiahParry** [reply]
I think that since `{arcgis}` doesn't list the `>= 1.0.0` it doesn't force you to do that....weird! Good to know!

**Reply by roimartins** [reply]
Also as you ask for suggestions or ideas , in my very humble (not properly working ) version of a  R Package to access to protected features I added following API parameters  in addition to "where" and "fields" . 

**spatial_layer :**  It was useful to have teh option to extract both spatial data or only data . If true 'returnGeometry' is set to True , if only the data is required will be set FALSE and return a data.frame
**n_records**:  When you want to do a quick exploration of the data without large download query 
**dist_values**: when you want to create a WHERE query sometimes you need to know unique values avaialble in the data ( e.g. years in the data ) 
**return_api**: in case you need to use teh API url in other analysis systems ( e.g. QGIS ) . I am working mainly with ArcPro but also significant  QGIS users , so always useful give that flexibility . 

In addition woudl be useful exploit other API options to minimise the amoun of data to be downloaded , e.g. the  intersect geometry option . 

```r
getWebServiceData = function ( web_service, layer_index , where_clause = "1=1", output_fields = "*", n_records = NULL, dist_values = FALSE,  spatial_layer = TRUE  , token_str = NULL, return_api = FALSE) {

[ ...]

  webservice_base$query = list(  where =  where_clause,                                    
                                 outFields = output_fields,                                
                                 resultRecordCount= n_records ,
                                 returnDistinctValues= dist_values ,
                                 returnGeometry = return_geom,                             
                                 f = format_out ,                                          ## set the  OUTPUT FORMAT
                                 token = token_str
                                 )

[...]


```

**Reply by JosiahParry** [reply]
- `arc_select()` provides the arguments `geometry`, `n_max`, and `filter_geom`.
- I'm not sure I follow what `dist_values` is supposed to do for you. 
- You can get the url from the `FeatureLayer` object by typing `flayer[["url"]]`

**Reply by JosiahParry** [reply]
If you wish to pass additional arguments to the API endpoint directly, you can utilize the `...` argument 

> `...` additional query parameters passed to the API.

**Reply by roimartins** [reply]
awesome! happy to drop to bin my cluncky function having these parameters available in arc_select ()  :) 

dis_values is boolean to set returnDistinctValue REST API parameter . I found it useful when you want to know distinct values available in a categorical field in your data   . For example,get the  distinct years avaiable in your data . Then you can use the WHERE SQL query to access a specific  years avaialbale in the data . As you mentioned in the documentation important to try limit the access to large datasets using WHERE clauses.

**Reply by JosiahParry** [reply]
In the mean time this can help address your objective: 

```r
furl <- paste0(
  "https://services3.arcgis.com/ZvidGQkLaDJxRSJ2/arcgis/rest/services/",
  "PLACES_LocalData_for_BetterHealth/FeatureServer/0"
)

flayer <- arc_open(furl)

x <- flayer |> 
  arc_select(
    where =  "TotalPopulation > 25000",
    fields = c("StateAbbr", "StateName"),
    returnDistinctValues = "true"
  )

Simple feature collection with 1676 features and 2 fields
Geometry type: POINT
Dimension:     XY
Bounding box:  xmin: -17590110 ymin: 2236138 xmax: -7657735 ymax: 9565120
Projected CRS: WGS 84 / Pseudo-Mercator
First 10 features:
   StateAbbr StateName                 geometry
1         AL   Alabama POINT (-9667788 3921451)
2         AL   Alabama POINT (-9516457 3843591)
3         AL   Alabama POINT (-9687220 3945138)
4         AL   Alabama POINT (-9662196 3965731)
5         AL   Alabama POINT (-9682951 4106540)
6         AL   Alabama POINT (-9507479 3663106)
7         AL   Alabama POINT (-9556321 3675367)
8         AL   Alabama POINT (-9758834 4140888)
9         AL   Alabama POINT (-9575200 4029973)
10        AL   Alabama POINT (-9663558 3956750)
```

**Reply by roimartins** [reply]
Thanks @JosiahParry , i know i udnerstand what you meant with '...' to pass additional  API arguments. That will do the job  and now definitevely very glad to decomission my script :) 

Impressive  job! I can see now the full AGE/AGOL integration with R  enviroment.

**Reply by JosiahParry** [reply]
Thanks for your feedback, @roimartins I was able to find two bugs that I need to fix today before i can submit to CRAN. It looks like `arc_select()` isn't respecting the `...` argument!

**Reply by JosiahParry** [reply]
both issues have fixed in the development versions thank you so much!!!

**Reply by roimartins** [reply]
thanks  to you , great to hear it was helpful . it was for me for sure. 

Just another feedback info may help: 

when I use the arc_select in a loop funciton to access the data for each year , it looks perform well the first year but then is giving the error below. When I run it manually then the data is extracted after a couple of attemps. 

 ```r 

for ( x in 2016:max(years)) {   
  print(x)    
  footprint_year =   arc_select( sar_f,  fields = fields_s,   where =  paste0 ( "year = ", x   ) )    
  print ( dim(fish_footprint_year))  
  L2_HA_footprint =   bind_rows( L2_HA_footprint ,  footprint_year  )   
}

> Warning message:
In parse_esri_json(httr2::resp_body_string(x)) : Status code: 500
Error: Error performing query operation

```

So the function receive  an sf with 0 records and the loops fail . Should be given some seconds between request  ?

**Reply by roimartins** [reply]
update on this: 

looping trhough last 3 years was extracting the data one year after the otehr with no interruptions. So may is this  related to internet connections ,other factors ?

**Reply by JosiahParry** [reply]
This is probably some rate limiting that is being enforced by the server that you are calling. 

Is there any reason why you need to perform this in a loop?

For example instead do: 

```r
years <- 2016:2023
max_year <- max(years)

arc_select(
  sar_f,
  fields = fields_s,
  where =  paste0("year >= 2016 and year <= ", max_year)
)
```

**Reply by roimartins** [reply]
@JosiahParry  may I am using in the wrong way the arc_open function but currently does not allow to use the function without a token . So if we want to use it with public services still require a token ? 


Even using the examples in the documentation rise the error with the missing token : 

 > furl <- paste0(
   "https://services3.arcgis.com/ZvidGQkLaDJxRSJ2/arcgis/rest/services/",
   "PLACES_LocalData_for_BetterHealth/FeatureServer/0"
 )

> arc_open(furl)

Error in `arc_open()`:
! Status code: 498
Error : Invalid token.
Run `rlang::last_trace()` to see where the error occurred.

**Reply by JosiahParry** [reply]
You probably have an expired token. You should restart your R session. If you provide a token it will always be checked first, even for public services. It's a little annoying. One thing we can do is to check if your token has expired and then remove it from your query with a warning. 

But, my guess is that you reopened a stale R session and tried to run this code. Try restarting your R session or removing your token with `unset_arc_token()`

**Reply by roimartins** [reply]
Thanks again, yes that was the issue.  The unset_arc_token() worked. 

In addition to your idea also may the error messsage could suggest potential solutions. e.g. Get a new token if the wervice is protectted or use the 'unset_arc_token()' if the web service is open ?

**Reply by roimartins** [reply]
I have been using also teh aptial predicates with arc_select 9) to get intersecgtionand works perfectly fine . Just I needed a bit of research to create the poitn as st_sfc () with the right projectio to be passed into the 'filter_geom' parameter. 

So , I sorted with :   haul_sel_geom =  st_point(c ( lon, lat )  )   |>  st_sfc() |> st_set_crs(4326)  

I like the idea that you dont have to pass the other parameters that are needed in the API GUI such : type of geometry and CRS of teh geometry . all is taken form teh sfc object. 

Great stuff!
